### PR TITLE
*: impl raftkv2 snapshot and write

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #12842 

What's Changed:

```commit-message
This PR implements snapshot and write for raftkv2.

Write are refactored to only use simple codec to reduce allocation
and parse overhead, it also makes code simpler.

Snapshot are refactored to return future instead of async function.
Otherwise it can't `Engine` requirement.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/server/raftkv2/mod.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    fn report_resolved(&self, _store_id: u64, _group_id: u64) {}
+
+    fn split(
+        &self,
+        _region_id: u64,
+        _region_epoch: kvproto::metapb::RegionEpoch,
+        _split_keys: Vec<Vec<u8>>,
+        _source: String,
+    ) -> futures::future::BoxFuture<'static, tikv_kv::Result<Vec<kvproto::metapb::Region>>> {
+        Box::pin(async move { Err(box_err!("raft split is not supported")) })
```
Comment (tonyxuqqi) [inline src/server/raftkv2/mod.rs:71]: I suppose split will be in another PR right? Maybe the error message could be "is not implemented yet."

File: src/server/raftkv2/mod.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        let res = self
+            .router
+            .store_router()
+            .send(region_id, msg)
+            .map_err(|e| tikv_kv::Error::from(raftstore_v2::Error::from(e)));
+        (Transform {
+            resp: CmdResStream::new(sub),
+            early_err: res.err(),
+        })
+        .inspect(move |ev| {
```
Comment (tonyxuqqi) [inline src/server/raftkv2/mod.rs:285]: where's the definition of the inspect function?  It seems not in either Stream or Transform

**Review by tonyxuqqi — APPROVED** [review]

File: src/server/raftkv2/mod.rs (hunk 1 above)
Comment (BusyJay) [reply src/server/raftkv2/mod.rs:71]: Yes, I just copied that from the definition of trait.

File: components/raftstore-v2/src/router/response_channel.rs [hunk 3]
```diff
@@ -203,13 +208,16 @@ impl<'a, Res> Future for WaitResult<'a, Res> {
...
-                    return Poll::Ready((*self.core.res.get()).take());
+                    return Poll::Ready((*self.sub.core.res.get()).take());
                 }
             }
         }
     }
 }
 
+unsafe impl<'a, Res: Send> Send for WaitResult<'a, Res> {}
+unsafe impl<'a, Res: Send> Sync for WaitResult<'a, Res> {}
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply components/raftstore-v2/src/router/response_channel.rs:219]: No, it doesn't need to. `Sync` is guaranteed by the definition of `WaitResult`.

File: src/server/raftkv2/mod.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        let data = encoder.encode();
+        let (ch, sub) = match on_applied {
+            Some(cb) => {
+                let (ch, sub) = CmdResChannel::with_callback(move |resp| {
+                    let mut res = if !resp.get_header().has_error() {
+                        Ok(())
+                    } else {
+                        Err(tikv_kv::Error::from(resp.get_header().get_error().clone()))
+                    };
+                    cb(&mut res);
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/server/raftkv2/mod.rs:264]: It's explained in the definition of the trait.

https://github.com/tikv/tikv/blob/77e4896507501059131e8951523d1050fb464b0f/components/tikv_kv/src/lib.rs#L336-L346

File: src/server/raftkv2/mod.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+};
+
+#[derive(Clone)]
+pub struct RaftExtensionImpl<EK: KvEngine, ER: RaftEngine> {
+    router: StoreRouter<EK, ER>,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> RaftExtension for RaftExtensionImpl<EK, ER> {
+    #[inline]
+    fn feed(&self, msg: RaftMessage, key_message: bool) {
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/server/raftkv2/mod.rs:32]: `log_error` is just an implementation details. Implementation also can choose to panic.

File: components/raftstore-v2/src/router/response_channel.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tabokie) [reply components/raftstore-v2/src/router/response_channel.rs:219]: It will catch warning no matter the implementation.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/raftstore-v2/src/router/response_channel.rs:219]: Can you provide steps to trigger the warnings?

[the remark this answers is not in the document]
Comment (tabokie) [reply components/raftstore-v2/src/router/response_channel.rs:219]: I haven't run [Rudra](https://github.com/sslab-gatech/Rudra) personally, but in the past the security team has alerted a similar [pattern](https://github.com/tikv/raft-engine/pull/146) in raft-engine.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/raftstore-v2/src/router/response_channel.rs:219]: The case you provided was probably the `Mutex` didn't seem to protect any data to Rudra as in Rust data is expected to put inside the `Mutex`. Here the situation is different. The implementation pattern is very common in lock-free data structs. I'm not sure whether Rudra can recognize it well, but if it doesn't then we may need to consider fixing the tools as it can also make a lot of false positive for many libraries in practice.

[the remark this answers is not in the document]
Comment (tabokie) [reply components/raftstore-v2/src/router/response_channel.rs:219]: Where is sync needed here? It seems all usage within this file can be changed to mutable reference.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/raftstore-v2/src/router/response_channel.rs]: It's not needed any more as now `BaseSubscriber` is borrowed.

**Review by tabokie — APPROVED** [review]
